delta-io/connectors
# Flink Delta Source PR 10 - get table schema from Delta Log.

This PR introduces a refactoring for Delta Source Builder.
The column names and types are no longer mandatory options for builder. 
Now user can define List of column names (Strings) that should be read from Delta table. Source connector will use Delta Log to get proper types. 
If no columns were defined by user, then source connector will read all columns from Delta Table. Schema discovery will be done using DeltaLog API.

PR Plan:
PR 5 - Option support in Bounded Mode - **DONE**
PR 6 - Continuous mode without monitoring for changes - **DONE**
PR 6.1 - Package Refactoring - **DONE**
PR 7 - Monitoring for changes in Continuous mode - **DONE**
PR 7.1 - More tests for PR 7 including IT case tests for Continuous mode. - **DONE**
PR 9 - Builder - **DONE**
**PR 10 - Builder refactoring to extract Schema from Delta Log.**
PR 10.1 - Tests for SourceUtils::buildSourceSchema, DeltaSourceBuilderBase::getSourceSchema and SnapshotSupplier implementations.
PR 11 - Partition Support using Delta log to identify partition columns.
PR 12 - Get BATCH_SIZE for Format builder from Source options - this value is currently hardcoded. Add validation for option names and values including using not applicable options for given mode via generic option method.

-- All functionalities are in place up to this point. The following PRs are extra tests, eventual bug fixes and code/javadoc adjustments. --

~~PR 11 - Annotate with @deprecated  Sink's builder methods that names starts from "with". Add new ones without "with" prefix and overload partitions with List argument for Sink.~~

PR 13 - Additional IT tests including all options and end to end tests using Source and Sink.

While @kristoffSC is on vacation
@tdas + @scottsand-db to review public API
Once Krzysztof is back
README / docs
QA
end-to-end integration tests - using Sink to write, Source to read, etc.
more examples / migrate examples from internal flink module to examples folder

TODOs and other improvements planed as next:

- Use index for keeping track of already processed paths in SnapshotProcessor - needs new Delta API.
- Do do not create snapshot when streaming only changes (startingVersion or startingTimestamp option used). - needs new Delta API since for startingTimestamp we still need to create a snapshot. There is no API like deltaLog.getChangesFromTimestamp(....)
- Refactor Test util classes to make them common for sink and source.
- Re-visit Delta Standalone APIs for when the source see's metadata/protocol actions. Update ActionProcessor class to handle MetaData and Protocol Actions
- Enhance DeltaLog.java with a new method similaiar `todeltaLog.getSnapshotForVersionAsOf()` but for streaming mode. Use it in Delta Source connector for `startingTimestamp` option. -> https://github.com/delta-io/connectors/pull/346
